Game/Resources/final_event/MainScript.as

```diff
@@ -131,14 +110,60 @@ class MainScript {
     void ReceiveMessage(Entity @sender, int signal) {
         switch (signal) {
             case 0: { // When monster has successfully eaten the player.
+                PrepareForFade();
                 phase = 4; // Lost phase.
                 print("Player: I'm losing.\n");
-                
                 break;
             }
         }
     }
 
+    void PrepareForFade() {
+        hub.renderManager.SetFogApply(true);
+        hub.renderManager.SetColorFilterApply(true);
+        hub.renderManager.SetFogColor(vec3(0, 0, 0));
+    }
+
+    bool DoFade(float deltaTime) {
+        // The idea of fading is that we start with fog to begin fading
+        // with distance. However, with fog we can generally see things
+        // close to us. To finish of the fade, a color filter approaching
+        // zero is used when the fog has darkened most of the view. The
+        // color filter applies to the entire screen uniformly which is
+        // why it's not used exclusively, but works nicely when fog has
+        // consumed most of the scene.
+
+        // Cube curve to get a density function that grows slowly at
+        // first and faster later to darken things close to us as well.
+        fogTimer += deltaTime;
+        float fogRatio = fogTimer / fogApexDuration;
+        fogRatio = fogRatio * fogRatio * fogRatio;
+
+        bool completed = false;
+        if (fogTimer >= fadeStartTime) {
+            fadeTimer += deltaTime;
+            float fadeRatio = fadeTimer / fadeApexDuration;
+            if (fadeRatio >= 1.0f) {
+                fadeRatio = 1.0f;
+                completed = true;
+            }
+            hub.renderManager.SetColorFilterColor(vec3(1.0f - fadeRatio, 1.0f - fadeRatio, 1.0f - fadeRatio));
+        }
+
+        hub.renderManager.SetFogDensity(fogRatio);
+
+        return completed;
+    }
+
+    void GameOver(bool won) {
+        phase = 1337;
```

Comment:
Really?

Reply:
Here we enter some state for when the game is over. I can't very well use one of the existing ones for this purpose. Still, we probably want to quit the game or something in which case a phase transition is not needed anyway. Since it's not specified how we want to end the game, I'm just entering an unused phase for now so that we don't do anything at all. Any suggestion as to how we should deal with the game being over?

Reply:
the concept is ok, the phase number is .......